src/main/scala/util/HeterogeneousBag.scala

```diff
@@ -2,17 +2,17 @@
 
 package freechips.rocketchip.util
 
-import Chisel._
-import chisel3.Record
+import chisel3._
+import chisel3.experimental.DataMirror
 import scala.collection.immutable.ListMap
 
 final case class HeterogeneousBag[T <: Data](elts: Seq[T]) extends Record with collection.IndexedSeq[T] {
-  def apply(x: Int) = elts(x)
+  val elements = ListMap(elts.zipWithIndex.map { case (n,i) => (i.toString, DataMirror.internal.chiselTypeClone(n)) }:_*)
```

Comment:
Since rocket-chip is on Scala 2.13 now, I would recommend using `VectorMap` instead of `ListMap`, it's **much** faster with large collection sizes.